PrefectHQ/prefect
# Fix `prefect dev start`

### Summary
As @abrookins noted in #7946, running `prefect dev start` no longer works.

### Expected behavior
Running `prefect dev start` without any flags should start auto-reloading instances of the Prefect Orion API, Prefect's Vue front-end, and a Prefect Agent. Editing code in Prefect code should restart Orion and the agent, and editing `orion-ui` code should restart the UI dev server.

### Current behavior
Running `prefect dev start` causes several errors:
- `AttributeError: 'ArgumentInfo' object has no attribute 'replace'` x 2
- `AttributeError: 'PrefectTyper' object has no attribute 'console'`

### Why is this happening?
There are two separate issues at play.

#### Typer default values 
`prefect dev start` calls one of Prefect's Typer CLI functions ([`prefect.cli.agent.start`](https://github.com/PrefectHQ/prefect/blob/d284bc6acd0bf4b1322db97fd028c64ac4565bb6/src/prefect/cli/agent.py#L41))directly from Python. However, `typer.Argument` and `typer.Option` default values only work if Typer calls the function. 

The end result: the agent runs, but crashes when it tries to use a Typer `ArgumentInfo` instance as a string.

#### Phantom console
`prefect.cli.agent.start` raises an error when trying to call `app.console.print`. `console` gets [added to `app`](https://github.com/PrefectHQ/prefect/blob/d284bc6acd0bf4b1322db97fd028c64ac4565bb6/src/prefect/cli/root.py#L74) when our Typer app starts. 

But `prefect dev start` spawns a subprocess that directly calls the agent's `start` function, so the Typer app never bootstraps and `console` is never added to `app`. This causes an `AttributeError` when `prefect.cli.agent.start` tries to write to the console.

### What this PR changes
- Adds a wrapper function to use when running the agent subprocess. The wrapper function does two things before running the agent:
    - Reads the signature of the agent start function and, where needed, extracts the correct default values from Typer's parameter helper objects. 
    - Attaches a [Rich](https://rich.readthedocs.io/en/stable/introduction.html) for the agent to use.

This fixes the issues in #7946.

#### But wait, there's more!
Fixing the first two errors revealed more trouble:
* We were using [watchfiles](https://pypi.org/project/watchfiles/) directly in the main process to manage agent restarts, and indirectly in a subprocess via Uvicorn's `--reload` flag. 
    * This seemed to prevent the Orion API from terminating. I had to force kill its process. It was fine on its own; when running `prefect dev start` with the `--no-agent` flag, Orion shut down as expected when I pressed Ctrl-C.
    * watchfiles uses Rust's [Notify library](https://github.com/notify-rs/notify), and others have seen deadlocks with it when [watching and unwatching the same file(s) from several threads](https://github.com/notify-rs/notify/issues/463). Not *exactly* the same as what I saw, but maybe related.
    * YMMV depending on OS and Python version, but I ran into this across multiple Python versions on MacOS.
* A [bug in watchfiles](https://github.com/samuelcolvin/watchfiles/issues/200) caused Rust `Already borrowed` errors to bubble up to Python when terminating `prefect agent start`, except on Python 3.11. 

#### Extra fixes
This PR also contains fixes for the watchfiles-related issues:
* Removed the `--reload` flag from Uvicorn's launch command and instead uses watchfiles directly to watch Prefect's source directory and restart the server when any files change. 
* Added exception handlers for the `Already borrowed` errors.
    * I'll look into fixing the error upstream. My Rust is rusty, but the problem looks straightforward.
* Added a `work_queues` parameter to both the `api` and `start` functions in `dev.py` to make it easy to connect the dev agent to different work queues if needed. Defaults to the `default` work queue.

### Example
* Clone the Prefect codebase and install it with `pip install -e ".[dev]"`.
* Run `prefect dev start` and notice you encounter the same errors reported in #7946.
* Checkout the `fix-prefect-dev-start` branch and run `prefect dev start` again. Notice that the errors no longer appear. 

### Steps to QA
* Run `prefect dev start`.
* Verify Orion is running be loading `http://localhost:4200` in a browser. 
* Verify the UI server works by loading the URL it shows on startup. 
* Verify the agent started and connected to the API by observing its console output.
* Edit files in the `src/prefect` directory and verify that Orion and the agent both restart. 
* Edit UI files in `orion-ui` and verify that the updates are pushed directly to the browser.

#### QA I did
* I ran `prefect dev start` and observed it working on both MacOS and Linux.
    * All services started up as expected.
    * Agent and Orion services restarted as expected when files were changed.
    * UI hot reloading worked as expected; changes to UI components were pushed straight to the brrowser.
    * I did not test on Windows as per [Prefect's docs on contributing](https://docs.prefect.io/contributing/overview/), dev tooling is not officially supported on Windows. Ensuring Windows compatibility for `prefect dev start` should only require minor changes if needed.

Closes #7946.

### Checklist


- [*] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [*] This pull request includes tests or only affects documentation. 
    - **I added tests covering the fixes for #7946, and a test that verifies `prefect dev start` tries to launch all three dev services. But the CLI's `dev` features had no tests before this, and could use more.**
- [*] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`
